include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td

```diff
@@ -206,6 +206,21 @@ def AddOfSelf : Pat <
   (MoveNameHint $old, (ShlPrimOp $x, (NativeCodeCall<"1">))),
   [(KnownWidth $x)]>;
 
+def minSize : NativeCodeCall<"$_builder.getI32IntegerAttr(std::min($0.getType().cast<FIRRTLBaseType>().getBitWidthOrSentinel(), $1.getType().cast<FIRRTLBaseType>().getBitWidthOrSentinel()))">;
+
+def NonIdentPad : Constraint<CPred<"$1.getValue() != $0.getType().cast<FIRRTLBaseType>().getBitWidthOrSentinel()">>;
+
+// add((zext a, n), (zext b, m)) -> zext(add(a, (zext b, m-n)), n)
+def AddOfZExtL : Pat <
+  (AddPrimOp:$old (PadPrimOp $a, $m), $b),
```

Comment:
Offhand this looks like it assumes unsigned integers -- I think pad will sign-extend SInt's, and I don't think this is safe there but haven't thought enough about it.

Reply:
narrowing is safe for both signed and unsigned and pad will zext or sext based on type

Reply:
I've also added sub.